### PR TITLE
Add ServerTrace and ClientTrace

### PR DESCRIPTION
**Pull request by svantreeck**
This adds the classes `ServerTrace` and `ClientTrace` and refactors TracingMiddleware and TracingHandler to take advantage of the functionality.

It also updates the project to VS2017 format from the pre-2017 project.json format (hopefully this is OK, but I didn't have the older tooling available).

### CODE REVIEW

File: zipkin4net-aspnetcore/Criteo.Profiling.Tracing.Middleware/TracingMiddleware.cs [hunk 1]
```diff
@@ -9,28 +10,24 @@ public static class TracingMiddleware
...
-                {
-                    await next.Invoke();
-                }
-                catch (System.Exception e)
+                else
                 {
-                    trace.Record(Annotations.Tag("error", e.Message));
-                    throw;
+                    // make the server trace a child of the passed in trace
+                    trace = trace.Child();
```
Comment (fedj) [inline zipkin4net-aspnetcore/Criteo.Profiling.Tracing.Middleware/TracingMiddleware.cs]: I think that it shouldn't be a Child of the received trace. A full span is composed of CS-SR-SS-CR annotations with the same spanId. If we call child here, then the span won't be complete as SR-SS would have another spanId

File: backup/zipkin4net-aspnetcore/Criteo.Profiling.Tracing.Middleware/Criteo.Profiling.Tracing.Middleware.csproj [hunk 2]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (fedj) [inline backup/zipkin4net-aspnetcore/Criteo.Profiling.Tracing.Middleware/Criteo.Profiling.Tracing.Middleware.csproj]: I doesn't seem to be a VS2017 csproj. Is it on purpose?

File: backup/zipkin4net/Criteo.Profiling.Tracing.UTest/Criteo.Profiling.Tracing.UTest.csproj [hunk 3]
```diff
@@ -0,0 +1,93 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (fedj) [inline backup/zipkin4net/Criteo.Profiling.Tracing.UTest/Criteo.Profiling.Tracing.UTest.csproj]: It doesn't seem to be a VS2017 csproj. Is it on purpose?

File: zipkin4net-example/common/common.csproj [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    <Copyright>Copyright 2017, Criteo</Copyright>
+    <VersionPrefix>0.1</VersionPrefix>
+    <Authors>Criteo</Authors>
+    <TargetFrameworks>net451;netstandard1.6</TargetFrameworks>
+    <AssemblyName>common</AssemblyName>
+    <PackageId>common</PackageId>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\zipkin4net\Criteo.Profiling.Tracing\Criteo.Profiling.Tracing.csproj" />
```
Comment (fedj) [inline zipkin4net-example/common/common.csproj]: I think we should directly use the nugets. IMO, it would be easier to detect problems with the nugets themselves. It also applies for backend and front-end projects.

File: zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        /// <param name="action"></param>
+        public static void TracedAction(Action action)
+        {
+            try
+            {
+                action();
+            }
+            catch (Exception ex)
+            {
+                Trace.Current?.Record(Annotations.Tag("error", ex.Message));
```
Comment (fedj) [inline zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs]: We're trying to stick with .Net v4.5 as many people don't have 4.6 installed. If I'm not wrong, '?.' syntactic sugar appeared in v4.6

File: zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        /// <param name="task"></param>
+        public static async Task TracedActionAsync(Task task)
+        {
+            try
+            {
+                await task;
+            }
+            catch (Exception ex)
+            {
+                Trace.Current?.Record(Annotations.Tag("error", ex.Message));
```
Comment (fedj) [inline zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs]: We're trying to stick with .Net v4.5 as many people don't have 4.6 installed. If I'm not wrong, '?.' syntactic sugar appeared in v4.6

File: zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        /// <param name="task"></param>
+        public static async Task<T> TracedActionAsync<T>(Task<T> task)
+        {
+            try
+            {
+                return await task;
+            }
+            catch (Exception ex)
+            {
+                Trace.Current?.Record(Annotations.Tag("error", ex.Message));
```
Comment (fedj) [inline zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs]: We're trying to stick with .Net v4.5 as many people don't have 4.6 installed. If I'm not wrong, '?.' syntactic sugar appeared in v4.6

File: backup/zipkin4net-aspnetcore/Criteo.Profiling.Tracing.Middleware/Criteo.Profiling.Tracing.Middleware.csproj [hunk 8]
```diff
@@ -1,54 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (fedj) [inline backup/zipkin4net-aspnetcore/Criteo.Profiling.Tracing.Middleware/Criteo.Profiling.Tracing.Middleware.csproj]: Why did you remove this csproj?

File: zipkin4net/Criteo.Profiling.Tracing/ClientTrace.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace Criteo.Profiling.Tracing
+{
+    public class ClientTrace : IDisposable
+    {
+        public Trace Trace { get; private set; }
+
+        public ClientTrace(string serviceName, string rpc)
+        {
+            Trace = Trace.Current.Child();
```
Comment (fedj) [inline zipkin4net/Criteo.Profiling.Tracing/ClientTrace.cs:16]: The child is already created in TracingHandler as we need to pass the context in the headers. There is an inconsistency here, you should choose to create it here or in TracingHandler but not both

**Review by fedj — CHANGES_REQUESTED** [review]
Thanks a lot for the PR! I think we should also modify .travis.yml as it won't compile anything now (it's using dotnet preview2 3131 and therefore will look for project.json files. I found this blog post to help you http://blog.wuc.me/2017/03/09/dotnet-core-travis.html change the version of dotnet for travis.
Your changes to go to 2017 csproj is very helpful and I think the  API changes are a brilliant idea. Thanks a lot for your work.

File: zipkin4net/Criteo.Profiling.Tracing/ClientTrace.cs (hunk 9 above)
Comment (svantreeck) [reply zipkin4net/Criteo.Profiling.Tracing/ClientTrace.cs:16]: Good catch, I'll fix this up

File: zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs (hunk 5 above)
Comment (svantreeck) [reply zipkin4net/Criteo.Profiling.Tracing/Utils/TraceHelper.cs]: This shouldn't end up being a problem...the `?.` operator is a feature of the complier/Roslyn and came with VS2015+ rather than being tied to a specific version of the CLR.  I can switch it to `if (x != null) {...}` if you'd like though.

**Comment by svantreeck** [comment]
I'm running into some issues with the nunit test adapter post VS2017 upgrade.  When in netcoreapp, the VS console isn't picking it up; however, after switching to .net 4.5, it builds locally but does not on the build server.  I'm going to investigate some of the workarounds mentioned at https://github.com/nunit/nunit3-vs-adapter/issues/300 to see if any will allow the build server and VS to work.

**Comment by svantreeck** [comment]
The nunit vs test adapter doesn't work with the new VS2017 project format entirely unless you target a non-dotnet core version.  As a result, I added .net 4.x to the test project to get it running again.  On the build server, because the new tools use msbuild, it is looking for .net 4.x templates; however, once targetting specifically netcoreapp on tests, it is still failing due to not being able to find any tests (most likely because of the issue above).

I don't have anything on my PC that run the project.json format, so I'm wondering if you'd be willing to cherry pick the Client/ServerTrace updates on this into the existing project.json structure so both builds and tests work.  Then, hopefully when nunit works entirely with the VS2017/net project structure, we can update it.

**Comment by fedj** [comment]
You can download a previous version of the framework [here](https://github.com/dotnet/core/blob/master/release-notes/download-archives/1.0.1-preview2-download.md). Just update your global.json to target 3133 and it should work.
I can cherry pick but it will therefore be my contribution (I would rather like to keep it as yours)

**Comment by svantreeck** [comment]
Unfortunately I don't have much time at the moment, so I'd probably prefer
if you could cherry pick the changes

**Comment by svantreeck** [comment]
I'll close this since you added #34

**Comment by fedj** [comment]
Thanks a lot for this work @svantreeck
